rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,700 @@
+    already supported) to each component they want to use.
+*   Individual components rely on unenforceable assumptions on how to interpret
+    the input data consistently.
+*   The complexity of adding new logical data representations (for example,
+    tf.SequenceExample) scales with the number of components.
+
+## The need for supporting new physical storage formats in TFX
+
+Two factors drive this need:
+
+*   TFX needs to offer users more choices of the storage format, e.g,
+    [Apache Parquet](https://parquet.apache.org/).
+*   TFX wants to be able to choose the optimal storage format based on user’s
+    workload, in a user-transparent manner. A unified I/O abstraction would make
+    it easier to support a new physical format in TFX, since one would not have
+    to understand every single TFX component in order to implement such support.
+
+## TFX interoperability with the rest of the world
+
+If we choose a commonly available and adopted exchange format as our in-memory
+representation, our users will be able to use TFX components with much less
+effort on data conversion. This aligns with TFX’s long term vision.
+
+# User Benefit
+
+## TFX End Users
+
+While this change is transparent to end users, it will facilitate the design and
+implementation of many user-facing features, for example:
+
+*   Columnar storage format in TFX.
+*   Structured training examples.
+
+## Individual TFX component users
+
+We use TFXIO to refer to the proposed I/O abstraction layer. All TFX components
+will start using TFXIO to ingest the data and have a unified way of representing
+the data. Individual TFX component users would be able to implement TFXIO for
+their own data formats / storage formats that are not supported by TFX. By
+design, any such implementation will be readily accessible by all TFX
+components.
+
+## TFX developers
+
+Developers working on TFX infrastructure will not have to understand the
+internals of each component any more in order to make changes to I/O and parsing
+(for example, adding support for a new storage format for the training
+examples).
+
+Developers working on TFX components would benefit from sharing common
+operations against the unified in-memory representation, or even higher-level
+computations. For instance, suppose that we implement a sketch-based algorithm
+to compute approximate heavy hitters over this in-memory representation. We can
+now share this implementation inside both TFDV and TFT for their top-K feature
+value computation.
+
+# Design Proposal
+
+This design proposes **a common in-memory data representation**, **a way to
+translate that into TF feedables** (np.ndarray or EagerTensors) and **a set of
+APIs** each component can use to get both.
+
+![alt_text](20191017-tfx-standardized-inputs/overview.png)
+
+## Common in-memory data representation
+
+[Apache Arrow](https://arrow.apache.org/) will be used as the common in-memory
+data representation. Beam-based TFX components will accept
```

Comment:
Does this depend on any new functionality in Apache Beam? (I tried searching for Arrow/Beam integration information, and in the codebase, and couldn't find much, but perhaps I was looking in the wrong place....)

Reply:
This does not depend on Apache Arrow integration with Apache Beam (which I think is at a very early stage..)

We simply pass around PCollection[RecordBatch].

Reply:
Deeper Arrow integration with Beam is still theoretical. I gave a talk about with some thoughts at the Beam Summit in September: https://s.apache.org/beam-arrow-vegas-2019

The tl;dr is you can do a lot with `PCollection[RecordBatch]` right now especially if you aren't doing anything with event times, windowing, etc.. but we have some more work to do (namely making schemas available across languages) before it's a real possibility.